zc_common/remote_resource/filters.py

```diff
@@ -1,55 +1,119 @@
 import re
 from distutils.util import strtobool
 
+from django.contrib.postgres.forms import SimpleArrayField
+from django.contrib.postgres.fields import ArrayField
 from django.db.models import BooleanField, FieldDoesNotExist
 from django.db.models.fields.related import ManyToManyField
+from django import forms
 from django.utils import six
 
 # DjangoFilterBackend was moved to django-filter and deprecated/moved from DRF in version 3.6
 try:
-    from rest_framework.filters import DjangoFilterBackend
+    from rest_framework.filters import DjangoFilterBackend, Filter
+    from rest_framework import filterset
 except ImportError:
     from django_filters.rest_framework import DjangoFilterBackend
+    from django_filters.rest_framework.filters import Filter
+    from django_filters.rest_framework import filterset
+
+
+class ArrayFilter(Filter):
+    field_class = SimpleArrayField
+
+    @property
+    def field(self):
+        # This property needs to be overriden because filters.Filter does not instantiate field_class with any
+        # args by default, and SimpleArrayField requires an argument indicating the type of each element in the array
+        self._field = self.field_class(forms.CharField(), required=False)
+        return self._field
+
+
+class JSONAPIFilterSet(filterset.FilterSet):
+    class Meta:
+        strict = True
+        filter_overrides = {
+            ArrayField: {
+                'filter_class': ArrayFilter,
+                'extra': lambda f: {
+                    'lookup_expr': 'contains',
+                }
+            }
+        }
 
 
 class JSONAPIFilterBackend(DjangoFilterBackend):
+    default_filter_set = JSONAPIFilterSet
+
+    # This method takes the filter query string (looks something like ?filter[xxx]=yyy) and parses into parameters
+    # that django_filters can interface with.
+    #
+    # Handles:
+    #   ?filter[id]=1
+    #   ?filter[id__in]=1,2,3
+    #   ?filter[price__gte]=100
+    #   ?filter[relatedobject__relatedobject]=1
+    #   ?filter[relatedobject__relatedobject__in]=1,2,3
+    def _parse_filter_string(self, queryset, filter_class, filter_string, filter_value):
+        filter_string_parts = filter_string.split('__')
+        if len(filter_string_parts) > 1:
+            field_name = '__'.join(filter_string_parts[:-1])
+        else:
+            field_name = filter_string_parts[0]
+
+        filterset_data = []
+
+        # Translates the 'id' in ?filter[id]= into the primary key identifier, e.g. 'pk'
+        if field_name == 'id':
+            primary_key = queryset.model._meta.pk.name
+            field_name = primary_key
+
+        try:
+            is_many_to_many_field = isinstance(getattr(queryset.model, filter_string).field, ManyToManyField)
+            if is_many_to_many_field:
+                filter_value = filter_value.split(',')
+        except AttributeError:
+            pass
+
+        try:
+            field_name = filter_string.split('__')[0]
```

Comment:
All my comments on this PR are hypothetical since I'm just trying to run it in my head, and I think there's a test that passes and would conflict with what I see here, but that said: does this line overwrite the value of `field_name` set above if there is a `__` in the filter_string?

This line appears to overwrite the block above where we do this:

```
filter_string_parts = filter_string.split('__')
        if len(filter_string_parts) > 1:
            field_name = '__'.join(filter_string_parts[:-1])
        else:
            field_name = filter_string_parts[0]
```

Reply:
You're right that we don't need this line and it overwrites the previous logic, but only if it the `if` case.

I think this would cause it to break in the case of:
`?filter[subscription_item__subscription__delivery_days__contains]=true`
where we're filtering on an ArrayField across multiple relationships.

I didn't have a test case to cover that situation, but I added one under `DeliveryItem`.